arangod/Scheduler/SupervisedScheduler.cpp

```diff
@@ -485,6 +492,59 @@ void SupervisedScheduler::shutdown() {
   }
 }
 
+Result SupervisedScheduler::detachThread(uint64_t* detachedThreads,
+                                         uint64_t* maximumDetachedThreads) {
+  std::lock_guard<std::mutex> guard(_mutex);
+  if (detachedThreads != nullptr) {
+    *detachedThreads = _numDetached;
+  }
+  if (maximumDetachedThreads != nullptr) {
+    *maximumDetachedThreads = _maxNumberDetachedThreads;
+  }
+  // First see if we have already reached the limit:
+  if (_numDetached >= _maxNumberDetachedThreads) {
+    return Result(TRI_ERROR_TOO_MANY_DETACHED_THREADS);
+  }
+
+  // Now we have access to the _workerStates and _detachedWorkerStates
+  // Let's first find ourselves in the _workerStates:
+  uint64_t myNumber = Thread::currentThreadNumber();
+  auto it = _workerStates.begin();
+  while (it != _workerStates.end()) {
+    if ((*it)->_thread->threadNumber() == myNumber) {
+      break;
+    }
+    ++it;
+  }
+  if (it == _workerStates.end()) {
+    return Result(TRI_ERROR_INTERNAL,
+                  "scheduler thread for detaching not found");
+  }
+  std::shared_ptr<WorkerState> state = std::move(*it);
+  _workerStates.erase(it);
+  // Since the thread is effectively taken out of the pool, decrease the
+  // number of workers.
+  --_numWorkers;
+  {
+    std::unique_lock<std::mutex> guard(state->_mutex);
+    state->_stop = true;  // We will be stopped after the current task is done
+                          // We know that we are working, so we do not
+                          // have to wake the thread.
```

Comment:
This lock should not be necessary since stop is an atomic.

Reply:
Same here (devel only).